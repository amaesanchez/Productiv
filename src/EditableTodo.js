import React, { useState } from "react";
import Todo from "./Todo";
import TodoForm from "./TodoForm";

/** Show editable todo item.
 *
 * Props
 * - todo
 * - update(): fn to call to update a todo
 * - remove(): fn to call to remove a todo
 *
 * States
 * - isEditing
 *
 * EditableTodoList -> EditableTodo -> { Todo, TodoForm }
 */

function EditableTodo({ todo, update, remove }) {
  const [isEditing, setIsEditing] = useState(false);

  /** Toggle if this is being edited */
  function toggleEdit() {
    // isEditing ? setIsEditing(false) : setIsEditing(true);
    setIsEditing(isEditing => !isEditing)
  }

  /** Call remove fn passed to this. */
  function handleDelete() {
    remove(todo.id);
  }

  /** Edit form saved; toggle isEditing and update in ancestor. */
  function save(todo) {
    toggleEdit();
    update(todo);
  }

  return (
    <div className="EditableTodo">
      {isEditing ? (
        <TodoForm initialFormData={todo} save={save} />
      ) : (
        <div className="mb-3">
          <div className="float-end text-sm-end">
            <button
              className="EditableTodo-toggle btn-link btn btn-sm text-decoration-none"
              onClick={toggleEdit}
            >
              Edit
            </button>
            <button
              className="EditableTodo-delBtn btn-link btn btn-sm text-danger text-decoration-none"
              onClick={handleDelete}
            >
              Del
            </button>
          </div>
          <Todo todo={todo} />
        </div>
      )}
    </div>
  );
}

export default EditableTodo;
