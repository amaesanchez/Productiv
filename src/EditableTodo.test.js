import { render, fireEvent } from "@testing-library/react";
import EditableTodo from "./EditableTodo";

const TODO_TEST = {
  id: 1,
  title: "Code!",
  description: "Write some code",
  priority: 2,
};

describe("EditableTodo", function () {
  it("renders without crashing", function () {
    render(<EditableTodo todo={TODO_TEST} />);
  });

  it("matches snapshot", function () {
    const { container } = render(<EditableTodo todo={TODO_TEST} />);
    expect(container).toMatchSnapshot();
  });

  it("Renders appropriate todo item in EditableTodoList area of DOM", function () {
    const { queryByText } = render(<EditableTodo todo={TODO_TEST} />);

    const todo = queryByText("Write some code");
    expect(todo).toBeInTheDocument();
  });

  it("Renders TodoForm when the edit button is clicked on an EditableTodo", function () {
    const { container, queryByText } = render(
      <EditableTodo todo={TODO_TEST} />
    );

    const todo = queryByText("Write some code");

    const editBtn = container.querySelector(".EditableTodo-toggle");
    fireEvent.click(editBtn);

    const editForm = container.querySelector(".NewTodoForm");

    expect(editForm).toBeInTheDocument();
    expect(todo).not.toBeInTheDocument();
  });

  // TODO: ask how to mock remove functionality (passed in as prop to EditableTodo)

  // it("Removes Todo when the delete button is clicked on an EditableTodo", function () {
  //   const { container, queryByText } = render(<EditableTodo todo={ TODO_TEST } />);

  //   const todo = queryByText("Write some code")
  //   expect(todo).toBeInTheDocument();

  //   const deleteBtn = container.querySelector(".EditableTodo-delBtn")
  //   fireEvent.click(deleteBtn)
  //   expect(todo).not.toBeInTheDocument();
  // })
});
