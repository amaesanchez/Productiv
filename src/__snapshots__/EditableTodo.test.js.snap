// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableTodo matches snapshot 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="float-end text-sm-end"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm text-decoration-none"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger text-decoration-none"
        >
          Del
        </button>
      </div>
      <div
        class="Todo"
      >
        <div>
          <b>
            Code!
          </b>
           
          <small>
            (priority: 
            2
            )
          </small>
        </div>
        <div>
          <small>
            Write some code
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`snapshot tests for stability matches snapshot 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="float-end text-sm-end"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm text-decoration-none"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger text-decoration-none"
        >
          Del
        </button>
      </div>
      <div
        class="Todo"
      >
        <div>
          <b>
            Code!
          </b>
           
          <small>
            (priority: 
            2
            )
          </small>
        </div>
        <div>
          <small>
            Write some code
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;
